File: solutions/cpp/space-age/1/space_age.cpp
#include "space_age.h"

namespace space_age {
    double space_age::on_mercury() const {
        return EarthReference / OrbitalPeriod::Mercury;
    }

    double space_age::on_venus() const {
        return EarthReference / OrbitalPeriod::Venus;
    }

    double space_age::on_earth() const {
        return EarthReference;
    }

    double space_age::on_mars() const {
        return EarthReference / OrbitalPeriod::Mars;
    }

    double space_age::on_jupiter() const {
        return EarthReference / OrbitalPeriod::Jupiter;
    }

    double space_age::on_saturn() const {
        return EarthReference / OrbitalPeriod::Saturn;
    }

    double space_age::on_uranus() const {
        return EarthReference / OrbitalPeriod::Uranus;
    }
    
    double space_age::on_neptune() const {
        return EarthReference / OrbitalPeriod::Neptune;
    }

    uint64_t space_age::seconds() const {
        return ConvertingFrom;
    }    

    space_age::space_age(const uint64_t& converting_from) 
        :
          ConvertingFrom { converting_from              },  
          EarthReference { converting_from / 31557600.0 }
    {

    }
}

File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

yearFromOrbPeriod :: Float -> Float -> Float
yearFromOrbPeriod seconds ratio = seconds / (31557600.0 * ratio) 

ageOn :: Planet -> Float -> Float
ageOn Mercury seconds = yearFromOrbPeriod seconds 0.2408467
ageOn Venus   seconds = yearFromOrbPeriod seconds 0.61519726
ageOn Earth   seconds = yearFromOrbPeriod seconds 1.0
ageOn Mars    seconds = yearFromOrbPeriod seconds 1.8808158
ageOn Jupiter seconds = yearFromOrbPeriod seconds 11.862615
ageOn Saturn  seconds = yearFromOrbPeriod seconds 29.447498
ageOn Uranus  seconds = yearFromOrbPeriod seconds 84.016846
ageOn Neptune seconds = yearFromOrbPeriod seconds 164.79132  